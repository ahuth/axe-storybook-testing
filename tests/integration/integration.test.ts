import {exec} from 'child_process';
import {it, expect} from 'vitest';

it('outputs accessibility violation information for the demo app', () => {
  expect.assertions(2);

  return new Promise<void>((done) => {
    exec('cd demo && npm run storybook:axeOnly', function (error, stdout) {
      const normalizedStdout = normalize(stdout);
      expect(error!.code).toEqual(1);
      expect(normalizedStdout).toMatchSnapshot();
      done();
    });
  });
}, 120_000);

it('filters the components to run', () => {
  expect.assertions(2);

  return new Promise<void>((done) => {
    exec(
      'cd demo && npm run storybook:axeOnly -- --pattern simple',
      function (error, stdout) {
        const normalizedStdout = normalize(stdout);
        expect(error!.code).toEqual(1);
        expect(normalizedStdout).toMatchSnapshot();
        done();
      },
    );
  });
}, 120_000);

it('fails only specific impact levels if specified', () => {
  expect.assertions(2);

  return new Promise<void>((done) => {
    exec(
      'cd demo && npm run storybook:axeOnly -- --failing-impact critical',
      function (error, stdout) {
        const normalizedStdout = normalize(stdout);
        expect(error!.code).toEqual(1);
        expect(normalizedStdout).toMatchSnapshot();
        done();
      },
    );
  });
}, 120_000);

/**
 * Remove items from a string that are specific to a test run or environment, such as timing
 * information and file-system paths. That way, we can snapshot test effectively.
 */
function normalize(input: string) {
  /** Test times reported by Mocha. For example, `(520ms)` or `(3s)` */
  const specTimePattern = /\s*\([\d.]+m?s\)/g;
  /** File system paths. For example, `/path/to/some/file */
  const cwdPattern = new RegExp(process.cwd(), 'g');
  /** Line numbers from stack trace paths. For example, `timers:20:55` */
  const lineNumbersPattern = /:\d+:\d+/g;
  /** storybook-start outputs an info dialog at the beginning this removes those lines */
  const storybookStartedFrame = /(│.*│\n|[─╭╮╰╯]+?)/g;
  /** webpack outputs a hash and build time */
  const webpackBuilt = /webpack built [\w\s]*\w\n/g;
  /** webpack outputs can vary from build to build */
  const webpackBuildTimeVariations =
    /info => Using cached manager|Opening `[\w/]+?` failed \(\d\): Device not configured\n/g;
  /** node show `process.` on some platforms and not others */
  const nodeProcessInternalsVariation = /process\.(.*?\(node:internal)/g;

  return input
    .replace(specTimePattern, '')
    .replace(cwdPattern, '')
    .replace(lineNumbersPattern, '')
    .replace(storybookStartedFrame, '')
    .replace(webpackBuilt, 'webpack built')
    .replace(webpackBuildTimeVariations, '')
    .replace(nodeProcessInternalsVariation, '$1');
}
