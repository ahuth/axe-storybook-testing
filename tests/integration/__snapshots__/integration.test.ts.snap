// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails only specific impact levels if specified 1`] = `
"$ axe-storybook --failing-impact critical
Serving up üçï static storybook build at: http://127.0.0.1:8000

[chromium] accessibility
  autoTitles
    ‚úî No Failures

  delays
    ‚úî Short Delay And Pass
    1) Short Delay And Fail
    ‚úî Medium Delay And Pass
    2) Medium Delay And Fail
    3) Medium Delay And Short Timeout Fail
    4) Long Delay And Timeout

  simple
    ‚úî No Failures
    5) Failure No Discernible Text
    ‚úî Failure Color Contrast
    6) Failure No Discernible Text And Invalid Role
    - Failure Color Contrast Skipped
    - Failure No Discernible Text And Invalid Role Skipped
    ‚úî Failure Color Contrast Disabled Rule
    7) Failure No Discernible Text And Invalid Role Disabled One Rule
    ‚úî Failure No Discernible Text And Invalid Role Disabled Rules


  7 passing
  2 pending
  7 failing

  1) delays
       Short Delay And Fail:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button id=\\"hi\\" role=\\"wut-the-wut\\">hello world</button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut
  

  2) delays
       Medium Delay And Fail:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button id=\\"hi\\" role=\\"wut-the-wut\\">hello world</button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut
  

  3) delays
       Medium Delay And Short Timeout Fail:
     Error: Timeout of 100ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers)
      at processTimers (node:internal/timers)

  4) delays
       Long Delay And Timeout:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers)
      at processTimers (node:internal/timers)

  5) simple
       Failure No Discernible Text:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  6) simple
       Failure No Discernible Text And Invalid Role:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut

     2. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  7) simple
       Failure No Discernible Text And Invalid Role Disabled One Rule:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  



info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
"
`;

exports[`filters the components to run 1`] = `
"$ axe-storybook --pattern simple
Serving up üçï static storybook build at: http://127.0.0.1:8000

[chromium] accessibility
  autoTitles
    - No Failures

  delays
    - Short Delay And Pass
    - Short Delay And Fail
    - Medium Delay And Pass
    - Medium Delay And Fail
    - Medium Delay And Short Timeout Fail
    - Long Delay And Timeout

  simple
    ‚úî No Failures
    1) Failure No Discernible Text
    2) Failure Color Contrast
    3) Failure No Discernible Text And Invalid Role
    - Failure Color Contrast Skipped
    - Failure No Discernible Text And Invalid Role Skipped
    ‚úî Failure Color Contrast Disabled Rule
    4) Failure No Discernible Text And Invalid Role Disabled One Rule
    ‚úî Failure No Discernible Text And Invalid Role Disabled Rules


  3 passing
  9 pending
  4 failing

  1) simple
       Failure No Discernible Text:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  2) simple
       Failure Color Contrast:
     Detected the following accessibility violations!

     1. color-contrast (Elements must have sufficient color contrast)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/color-contrast?application=axeAPI.

        Check these nodes:

        - html: <button style=\\"background-color: red; color: hotpink;\\">hello world</button>
          summary: Fix any of the following:
                     Element has insufficient color contrast of 1.51 (foreground color: #ff69b4, background color: #ff0000, font size: 10.0pt (13.3333px), font weight: normal). Expected contrast ratio of 4.5:1
  

  3) simple
       Failure No Discernible Text And Invalid Role:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut

     2. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  4) simple
       Failure No Discernible Text And Invalid Role Disabled One Rule:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  



info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
"
`;

exports[`outputs accessibility violation information for the demo app 1`] = `
"$ axe-storybook
Serving up üçï static storybook build at: http://127.0.0.1:8000

[chromium] accessibility
  autoTitles
    ‚úî No Failures

  delays
    ‚úî Short Delay And Pass
    1) Short Delay And Fail
    ‚úî Medium Delay And Pass
    2) Medium Delay And Fail
    3) Medium Delay And Short Timeout Fail
    4) Long Delay And Timeout

  simple
    ‚úî No Failures
    5) Failure No Discernible Text
    6) Failure Color Contrast
    7) Failure No Discernible Text And Invalid Role
    - Failure Color Contrast Skipped
    - Failure No Discernible Text And Invalid Role Skipped
    ‚úî Failure Color Contrast Disabled Rule
    8) Failure No Discernible Text And Invalid Role Disabled One Rule
    ‚úî Failure No Discernible Text And Invalid Role Disabled Rules


  6 passing
  2 pending
  8 failing

  1) delays
       Short Delay And Fail:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button id=\\"hi\\" role=\\"wut-the-wut\\">hello world</button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut
  

  2) delays
       Medium Delay And Fail:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button id=\\"hi\\" role=\\"wut-the-wut\\">hello world</button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut
  

  3) delays
       Medium Delay And Short Timeout Fail:
     Error: Timeout of 100ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers)
      at processTimers (node:internal/timers)

  4) delays
       Long Delay And Timeout:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure \\"done()\\" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (node:internal/timers)
      at processTimers (node:internal/timers)

  5) simple
       Failure No Discernible Text:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  6) simple
       Failure Color Contrast:
     Detected the following accessibility violations!

     1. color-contrast (Elements must have sufficient color contrast)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/color-contrast?application=axeAPI.

        Check these nodes:

        - html: <button style=\\"background-color: red; color: hotpink;\\">hello world</button>
          summary: Fix any of the following:
                     Element has insufficient color contrast of 1.51 (foreground color: #ff69b4, background color: #ff0000, font size: 10.0pt (13.3333px), font weight: normal). Expected contrast ratio of 4.5:1
  

  7) simple
       Failure No Discernible Text And Invalid Role:
     Detected the following accessibility violations!

     1. aria-roles (ARIA roles used must conform to valid values)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/aria-roles?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix all of the following:
                     Role must be one of the valid ARIA roles: wut-the-wut

     2. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  

  8) simple
       Failure No Discernible Text And Invalid Role Disabled One Rule:
     Detected the following accessibility violations!

     1. button-name (Buttons must have discernible text)

        For more info, visit https://dequeuniversity.com/rules/axe/4.4/button-name?application=axeAPI.

        Check these nodes:

        - html: <button role=\\"wut-the-wut\\"></button>
          summary: Fix any of the following:
                     Element does not have inner text that is visible to screen readers
                     aria-label attribute does not exist or is empty
                     aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                     Element has no title attribute
                     Element's default semantics were not overridden with role=\\"none\\" or role=\\"presentation\\"
  



info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
"
`;
